### PR TITLE
Add lots of commands and improvements.

### LINKED ISSUE #16: Use correct "committer" on merge
Currently all PRs merged by Jarvis look like they were merged by @jordansissel.  That is cool and all, but any problems should be pointed at the person doing the merge (ie telling jarvis to merge), but this information is lost.

### PR DESCRIPTION
**Pull request by jordansissel**
This is another large-sweep as part of the refactor/rewrite of Jarvis.

Major changes:

* Clamp delegation now supports passing flag values that come from the request or Lita's environment (see the 'merge' command fancy_route example)
* Implemented merge and cla commands.
* Added user profile settings via `user set [key] [value]` and `user set [key]`. This is used for telling Lita what your git name+email are for the purposes of merging. (Fixes #16)
* Much more logging is captured during merge execution and is published only on error. At this time, I think _all_ output from the merge command, on failure, will reach the user (instead of the console output!) for easier debugging.
* Verified 'restart' command works.
* Implemented ping command which also shows last-successful heartbeat time.
* I tried to implement heartbeat but Lita will ignore its own messages :(
* Added merge and publish commands.

### CODE REVIEW

File: lib/jarvis/clamp_delegate.rb [hunk 1]
```diff
@@ -13,19 +13,42 @@ module Jarvis module ClampDelegate
...
+        # prepend flags to the args
+        # The `:flags` value is expected to be a Hash of "--flag-name" => lambda { |request| ... }
+        # The idea is that there can be automatic flags based on the
+        # circumstances of the request.
+        options[:flags].each do |key, value_proc|
+          value = value_proc.call(request)
+          args.unshift(key, value) if value
+        end
+      end
+
```
Comment (ph) [inline lib/jarvis/clamp_delegate.rb]: I see 4 contexts of execution in this method:
- validate the options
- generating a command
- inject
- execute

I think we could split things up like this, a command class:

Command.new(command_class, options)
Command.valid?
Command.command_method
Command.request

inject_method(cmd.command_method, request)
execute

File: lib/jarvis/commands/cla.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    if cla.ok?
+      puts I18n.t("lita.handlers.jarvis.cla.success", :project => "#{user}/#{project}", :pr => pr, :message => cla.message)
+    else
+      puts I18n.t("lita.handlers.jarvis.cla.failure", :project => "#{user}/#{project}", :pr => pr, :message => cla.message)
+      return 1
+    end
+  rescue => e
+    puts "An error occurred: #{e.class} - #{e}\n" + e.backtrace.join("\n")
+  end
+end end end
```
Comment (ph) [inline lib/jarvis/commands/cla.rb]: This make it easy to add command ! :heart:

File: lib/jarvis/commands/merge.rb [hunk 3]
```diff
@@ -1,15 +1,127 @@
...
+    logs = []
+    logger = ::Cabin::Channel.new
+    logger.subscribe(logs)
+    logger.subscribe(STDOUT) if STDOUT.tty?
+    logger.level = :debug
+    logger[:context] = "#{pr.project}\##{pr.number}"
+
+    if workdir.nil?
+      self.workdir = Stud::Temporary.pathname
+      defer.do { FileUtils.rm_r(workdir, :secure => true) }
```
Comment (ph) [inline lib/jarvis/commands/merge.rb]: Smell like golang was here

File: lib/jarvis/commands/merge.rb [hunk 4]
```diff
@@ -1,15 +1,127 @@
...
 
   def execute
-    puts "Merging: #{url} into #{branches_list}"
+    defer = ::Jarvis::Defer.new
+    logs = []
+    logger = ::Cabin::Channel.new
+    logger.subscribe(logs)
+    logger.subscribe(STDOUT) if STDOUT.tty?
+    logger.level = :debug
+    logger[:context] = "#{pr.project}\##{pr.number}"
```
Comment (ph) [inline lib/jarvis/commands/merge.rb]: maybe encapsulate that into a `logger` method?

File: lib/jarvis/commands/merge.rb [hunk 5]
```diff
@@ -1,15 +1,127 @@
...
+    # TODO(sissel): Set labels on PRs
+    puts I18n.t("lita.handlers.jarvis.merge success", organization: pr.organization, project: pr.project, number: pr.number, branches: branches.join(","))
   rescue => e
-    puts "An error occurred: #{e.class} - #{e}\n" + e.backtrace.join("\n")
+    puts I18n.t("lita.handlers.jarvis.exception", :exception => e.class, :message => e.to_s, :stacktrace => e.backtrace.join("\n"), :command => "merge", :logs => logs.join("\n"))
+  ensure
+    defer.run
   end
+
 end end end
```
Comment (ph) [inline lib/jarvis/commands/merge.rb]: This could be split up into small methods, but let go forward with it.

File: lib/jarvis/defer.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  def do(&block)
+    @deferred << block
+  end
+
+  def run
+    @deferred.each do |block|
+      block.call
+    end
+  end
+end end
```
Comment (ph) [inline lib/jarvis/defer.rb]: We could probably run the defer call with concurrent ruby future? I don't think the order of execution is really important in that context.

File: lib/jarvis/exec.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      Open4::popen4(*wrapped)
+    else
+      Open4::popen4(*args)
+    end
+    stdin.close
+    logger.pipe(stdout => :info, stderr => :error)
+    _, status = Process::waitpid2(pid)
+    raise SubprocessFailure, "subprocess failed with code #{status.exitstatus}" unless status.success?
+  end
+end
```
Comment (ph) [inline lib/jarvis/exec.rb]: I will probably add a rbenv shim for that. I don't want to go back into rvm :D

File: lib/jarvis/fetch.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      raise DownloadFailure, "Got HTTP #{response.status} when fetching #{url}"
+    end
+
+    file.write(response.body)
+    file.flush
+    file.rewind
+
+    file
+  end
+end end
```
Comment (ph) [inline lib/jarvis/fetch.rb]: Later we do stream download.

File: lib/lita/handlers/heartbeat.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      end
+
+      def ping(response)
+        response.reply(I18n.t("lita.handlers.jarvis.ping response", :last_heartbeat => LastHeartBeat.get.iso8601))
+      end
+
+      Lita.register_handler(self)
+    end
+  end
+end
```
Comment (ph) [inline lib/lita/handlers/heartbeat.rb]: I see a lot of `puts` call, should we use a logger instead to stdout?

File: lib/jarvis/commands/merge.rb (hunk 3 above)
Comment (jordansissel) [inline lib/jarvis/commands/merge.rb]: hehe :P

File: lib/jarvis/defer.rb (hunk 6 above)
Comment (jordansissel) [inline lib/jarvis/defer.rb]: You mean having run() just schedule some futures instead of calling each individually? I like that idea.

Comment (ph) [inline lib/jarvis/defer.rb]: Yeah, you can always check for their execution status,

http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/Future.html

File: lib/jarvis/commands/publish.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      context.clear()
+      git.reset
+      git.clean(force: true)
+    end
+
+    puts I18n.t("lita.handlers.jarvis.publish success", organization: project.organization, project: project.name, branches: branches.join(", "))
+  rescue => e
+    puts I18n.t("lita.handlers.jarvis.exception", :exception => e.class, :message => e.to_s, :stacktrace => e.backtrace.join("\n"), :command => "merge", :logs => logs.collect { |l| l[:message] }.join("\n"))
+  end
+end end end
```
Comment (ph) [inline lib/jarvis/commands/publish.rb]: I think the publish command should do this scenario:

**pseudo code**
- pull branch
- get sha1
- ask jenkins for build status for this project and this sha1 **we should be able to skip that, for core v1 gem or when the build is toast**
- bundle exec rake vendor
- bundle exec rake publish_gem

the build need to be run under jruby to make sure the gem has the right platform.
I believe the only dependency for this process should be the `logstash-devutils` gem

Comment (ph) [inline lib/jarvis/commands/publish.rb]: I am checking this library https://github.com/arangamani/jenkins_api_client, the commit `sha1` should be available, we can see it in the job output log. We can probably filter the project by name and compare their sha1 with the cloned code.

Comment (ph) [inline lib/jarvis/commands/publish.rb]: To note: installing `dependencies/running` the spec is the slowest part of the publish and this is also where we see most of the errors.

File: lib/jarvis/clamp_delegate.rb (hunk 1 above)
Comment (jordansissel) [inline lib/jarvis/clamp_delegate.rb]: Filed as https://github.com/elastic/jarvis/issues/41

File: lib/jarvis/commands/merge.rb (hunk 4 above)
Comment (jordansissel) [inline lib/jarvis/commands/merge.rb]: https://github.com/elastic/jarvis/issues/42

File: lib/jarvis/commands/merge.rb (hunk 5 above)
Comment (jordansissel) [inline lib/jarvis/commands/merge.rb]: https://github.com/elastic/jarvis/issues/43

File: lib/jarvis/defer.rb (hunk 6 above)
Comment (jordansissel) [inline lib/jarvis/defer.rb]: https://github.com/elastic/jarvis/issues/44

File: lib/jarvis/exec.rb (hunk 7 above)
Comment (jordansissel) [inline lib/jarvis/exec.rb]: haha, fair ;P

File: lib/lita/handlers/heartbeat.rb (hunk 9 above)
Comment (jordansissel) [inline lib/lita/handlers/heartbeat.rb]: Yeah actually I"m going to delete the heartbeat stuff since Lita doesn't see messages from itself (by design)

Comment (jordansissel) [inline lib/lita/handlers/heartbeat.rb]: 2323cd9 removed the old heartbeat stuff

File: lib/jarvis/commands/publish.rb (hunk 10 above)
Comment (jordansissel) [inline lib/jarvis/commands/publish.rb]: https://github.com/elastic/jarvis/issues/45

**Comment by jordansissel** [comment]
Discussed over Zoom w/ @ph - I'm going to merge this as-is, and I've created issues for each major point to resolve. This will let us pair up on it easier and have smaller code reviews to do in the future.

**Pull request merged**